File: web-api/providers/volcengine.mjs
import crypto from "node:crypto";
import { writeApiLog } from "../core/database.mjs";
import { decryptSecret } from "../core/crypto.mjs";
import { getAccountSecretRecord } from "../repositories/accounts.mjs";

function encode(value) {
  return encodeURIComponent(String(value)).replace(/[!'()*]/g, (c) => `%${c.charCodeAt(0).toString(16).toUpperCase()}`);
}

function queryText(params) {
  return Object.entries(params)
    .filter(([, value]) => value !== undefined && value !== null)
    .map(([key, value]) => [encode(key), encode(value)])
    .sort((a, b) => a[0].localeCompare(b[0]) || a[1].localeCompare(b[1]))
    .map(([key, value]) => `${key}=${value}`)
    .join("&");
}

function account(accountId) {
  const row = getAccountSecretRecord(accountId);
  if (!row) throw new Error("云账号不存在");
  if (!row.enabled) throw new Error("云账号已停用");
  if (row.cloud_type !== "volcengine") throw new Error("当前账号不是火山引擎账号");
  return row;
}

function signingKey(row, service, region, date) {
  const secret = decryptSecret(row.secret_ciphertext);
  const dateKey = crypto.createHmac("sha256", secret).update(date).digest();
  const regionKey = crypto.createHmac("sha256", dateKey).update(region).digest();
  const serviceKey = crypto.createHmac("sha256", regionKey).update(service).digest();
  return crypto.createHmac("sha256", serviceKey).update("request").digest();
}

export async function request(accountId, service, version, action, params = {}, region = "cn-beijing") {
  const row = account(accountId);
  const selectedRegion = String(region || row.region_id || "cn-beijing");
  const host = "open.volcengineapi.com";
  const requestParams = { ...params, Action: action, Version: version };
  const query = queryText(requestParams);
  const dateTime = new Date().toISOString().replace(/[-:]|\.\d{3}/g, "");
  const date = dateTime.slice(0, 8);
  const canonicalHeaders = `x-date:${dateTime}\n`;
  const signedHeaders = "x-date";
  const canonicalRequest = `GET\n/\n${query}\n${canonicalHeaders}\n${signedHeaders}\n${crypto.createHash("sha256").update("").digest("hex")}`;
  const credentialScope = `${date}/${selectedRegion}/${service}/request`;
  const stringToSign = `HMAC-SHA256\n${dateTime}\n${credentialScope}\n${crypto.createHash("sha256").update(canonicalRequest).digest("hex")}`;
  const signature = crypto.createHmac("sha256", signingKey(row, service, selectedRegion, date)).update(stringToSign).digest("hex");
  const authorization = `HMAC-SHA256 Credential=${row.access_key_id}/${credentialScope}, SignedHeaders=${signedHeaders}, Signature=${signature}`;
  const response = await fetch(`https://${host}/?${query}`, { headers: { "X-Date": dateTime, Authorization: authorization } });
  const data = await response.json().catch(() => ({}));
  if (!response.ok || data?.ResponseMetadata?.Error || data?.Error) {
    const apiError = data?.ResponseMetadata?.Error || data?.Error || {};
    const message = apiError.Message || apiError.Code || data?.Message || `火山引擎 ${response.status}`;
    writeApiLog(accountId, host, action, requestParams, data, "失败", message);
    throw new Error(message);
  }
  writeApiLog(accountId, host, action, requestParams, data, "成功");
  return data?.Result ?? data;
}

export async function jsonRequest(accountId, service, version, action, payload = {}, region = "cn-beijing") {
  const row = account(accountId);
  const selectedRegion = String(region || row.region_id || "cn-beijing");
  const host = `${service}.volcengineapi.com`;
  const requestParams = { Action: action, Version: version };
  const query = queryText(requestParams);
  const body = JSON.stringify(payload);
  const dateTime = new Date().toISOString().replace(/[-:]|\.\d{3}/g, "");
  const date = dateTime.slice(0, 8);
  const payloadHash = crypto.createHash("sha256").update(body).digest("hex");
  const canonicalHeaders = `x-date:${dateTime}\n`;
  const signedHeaders = "x-date";
  const canonicalRequest = `POST\n/\n${query}\n${canonicalHeaders}\n${signedHeaders}\n${payloadHash}`;
  const credentialScope = `${date}/${selectedRegion}/${service}/request`;
  const stringToSign = `HMAC-SHA256\n${dateTime}\n${credentialScope}\n${crypto.createHash("sha256").update(canonicalRequest).digest("hex")}`;
  const signature = crypto.createHmac("sha256", signingKey(row, service, selectedRegion, date)).update(stringToSign).digest("hex");
  const authorization = `HMAC-SHA256 Credential=${row.access_key_id}/${credentialScope}, SignedHeaders=${signedHeaders}, Signature=${signature}`;
  const response = await fetch(`https://${host}/?${query}`, { method: "POST", headers: { Accept: "application/json", "Content-Type": "application/json", "X-Date": dateTime, Authorization: authorization }, body });
  const text = await response.text();
  let data;
  try { data = JSON.parse(text); } catch { data = { Message: text || `火山引擎 ${response.status}` }; }
  if (!response.ok || data?.ResponseMetadata?.Error || data?.Error) {
    const apiError = data?.ResponseMetadata?.Error || data?.Error || {};
    const message = apiError.Message || apiError.Code || data?.Message || `火山引擎 ${response.status}`;
    writeApiLog(accountId, host, action, requestParams, data, "失败", message);
    throw new Error(message);
  }
  writeApiLog(accountId, host, action, requestParams, data, "成功");
  return data?.Result ?? data;
}

function xmlText(xml, tag) { const match = String(xml).match(new RegExp(`<${tag}(?:\\s[^>]*)?>([\\s\\S]*?)</${tag}>`, "i")); return match ? match[1].replace(/<!\[CDATA\[([\s\S]*?)\]\]>/g, "$1").trim() : ""; }
function xmlBlocks(xml, tag) { return [...String(xml).matchAll(new RegExp(`<${tag}(?:\\s[^>]*)?>([\\s\\S]*?)</${tag}>`, "gi"))].map((match) => match[1]); }

export async function tosBuckets(accountId) {
  const row = account(accountId);
  const region = String(row.region_id || "cn-beijing");
  const host = `tos-${region}.volces.com`;
  const dateTime = new Date().toISOString().replace(/[-:]|\.\d{3}/g, "");
  const date = dateTime.slice(0, 8);
  const payloadHash = crypto.createHash("sha256").update("").digest("hex");
  const canonicalHeaders = `host:${host}\nx-tos-content-sha256:${payloadHash}\nx-tos-date:${dateTime}\n`;
  const signedHeaders = "host;x-tos-content-sha256;x-tos-date";
  const canonicalRequest = `GET\n/\n\n${canonicalHeaders}\n${signedHeaders}\n${payloadHash}`;
  const credentialScope = `${date}/${region}/tos/request`;
  const stringToSign = `TOS4-HMAC-SHA256\n${dateTime}\n${credentialScope}\n${crypto.createHash("sha256").update(canonicalRequest).digest("hex")}`;
  const signature = crypto.createHmac("sha256", signingKey(row, "tos", region, date)).update(stringToSign).digest("hex");
  const authorization = `TOS4-HMAC-SHA256 Credential=${row.access_key_id}/${credentialScope}, SignedHeaders=${signedHeaders}, Signature=${signature}`;
  const response = await fetch(`https://${host}/`, { headers: { Host: host, "X-Tos-Date": dateTime, "X-Tos-Content-Sha256": payloadHash, Authorization: authorization } });
  const xml = await response.text();
  if (!response.ok) {
    const message = xmlText(xml, "Message") || xmlText(xml, "Code") || `TOS ${response.status}`;
    writeApiLog(accountId, host, "ListBuckets", {}, { body: xml }, "失败", message);
    throw new Error(message);
  }
  let buckets = [];
  try { const data = JSON.parse(xml); buckets = Array.isArray(data?.Buckets) ? data.Buckets : (data?.Buckets?.Bucket || data?.Bucket || []); }
  catch { buckets = xmlBlocks(xml, "Bucket").map((block) => ({ Name: xmlText(block, "Name"), Location: xmlText(block, "Location"), CreationDate: xmlText(block, "CreationDate") })); }
  const items = buckets.map((bucket) => { const name = bucket.Name || bucket.BucketName || ""; const location = bucket.Location || bucket.Region || region; return { Name: name, Location: location, CreationDate: bucket.CreationDate || "", StorageClass: "Standard", ExtranetEndpoint: `${name}.tos-${location}.volces.com`, Acl: "private" }; }).filter((item) => item.Name);
  writeApiLog(accountId, host, "ListBuckets", {}, { count: items.length }, "成功");
  return items;
}

function arr(data, path) {
  let value = data;
  for (const key of path) value = value?.[key];
  if (!value) return [];
  return Array.isArray(value) ? value : [value];
}

function instance(item, region) { return { ...item, InstanceId: item.InstanceId || item.InstanceID, InstanceName: item.InstanceName || item.InstanceId || item.InstanceID, InstanceStatus: item.Status || item.InstanceStatus, Status: item.Status || item.InstanceStatus, PublicIpAddress: item.PublicIpAddress || item.PublicIpAddresses?.[0] || item.EipAddress || "", _region_id: region }; }
function rds(item, region) { return { ...item, DBInstanceId: item.DBInstanceId || item.InstanceId || item.InstanceID, DBInstanceDescription: item.DBInstanceName || item.InstanceName || item.DBInstanceId || item.InstanceId, DBInstanceStatus: item.Status || item.DBInstanceStatus, Engine: item.Engine || "MySQL", _region_id: region }; }
function swas(item, region) { return { ...item, InstanceId: item.InstanceId || item.InstanceID, InstanceName: item.InstanceName || item.Name || item.InstanceId || item.InstanceID, InstanceStatus: item.Status || item.InstanceStatus, Status: item.Status || item.InstanceStatus, PublicIpAddress: item.PublicIpAddress || item.PublicIp || item.PublicIpAddresses?.[0] || item.EipAddress || "", _region_id: region }; }
function redis(item, region) { const instanceId = item.InstanceId || item.InstanceID || item.DBInstanceId || item.RedisInstanceId; return { ...item, KVStoreInstanceId: instanceId, InstanceId: instanceId, InstanceName: item.InstanceName || item.DBInstanceName || item.Name || instanceId, InstanceStatus: item.Status || item.InstanceStatus, DBInstanceStatus: item.Status || item.InstanceStatus, EngineVersion: item.EngineVersion || item.RedisVersion || "Redis", _region_id: region }; }
function edgeDomain(item) { const domain = item.DomainName || item.Domain || item.Name || ""; return { ...item, SiteId: item.DomainId || item.DomainID || domain, SiteName: domain, DomainName: domain, Status: item.Status || item.DomainStatus || "", AccessType: item.ServiceType || item.BusinessType || "CDN", Coverage: item.Area || item.Scope || "", PlanName: item.Plan || item.ProductType || "" }; }
function dnsZone(item) { const expiresAt = Number(item.ExpiredTime || 0); return { ...item, DomainName: item.ZoneName || "", DomainStatus: "正常", RegistrationDate: item.CreatedAt || "", ExpirationDate: expiresAt ? new Date(expiresAt < 1e12 ? expiresAt * 1000 : expiresAt).toISOString() : "", RecordCount: item.RecordCount || 0 }; }

export async function resources(accountId, type) {
  const row = account(accountId);
  const region = String(row.region_id || "cn-beijing");
  const errors = [];
  try {
    if (type === "ecs") { const data = await request(accountId, "ecs", "2020-04-01", "DescribeInstances", { PageSize: 100, PageNumber: 1 }, region); const items = arr(data, ["Instances"]).length ? arr(data, ["Instances"]) : arr(data, ["Instances", "Instance"]); return { resource_type: type, items: items.map((item) => instance(item, region)), errors, fetched_at: Date.now() }; }
    if (type === "oss") return { resource_type: type, items: await tosBuckets(accountId), errors, fetched_at: Date.now() };
    if (type === "domain") { const data = await jsonRequest(accountId, "dns", "2018-08-01", "ListZones", { PageSize: 100, PageNumber: 1 }, region); return { resource_type: type, items: arr(data, ["Zones"]).map(dnsZone), errors, fetched_at: Date.now() }; }
    const definitions = { rds: ["rds_mysql", "2018-01-01", "DescribeDBInstances", rds, "DBInstances", "Items"], swas: ["lighthouse", "2020-04-01", "DescribeInstances", swas, "Instances", "InstanceSet"], redis: ["Redis", "2020-12-07", "DescribeDBInstances", redis, "DBInstances", "Items"], esa: ["cdn", "2021-03-01", "ListCdnDomains", edgeDomain, "Domains", "DomainList"] };
    const definition = definitions[type];
    if (!definition) return { resource_type: type, items: [], errors: [`火山引擎暂未接入 ${type} 资源`], fetched_at: Date.now() };
    const [service, version, action, normalize, firstKey, fallbackKey] = definition;
    const data = await request(accountId, service, version, action, { PageSize: 100, PageNumber: 1 }, region);
    const items = arr(data, [firstKey]).length ? arr(data, [firstKey]) : arr(data, [fallbackKey]);
    return { resource_type: type, items: items.map((item) => normalize(item, region)), errors, fetched_at: Date.now() };
  } catch (error) { errors.push(error.message); return { resource_type: type, items: [], errors, fetched_at: Date.now() }; }
}

File: src-tauri/src/cloud/volcengine.rs
use crate::{array_at, string_params, volc_region_id, write_api_log, xml_blocks, xml_text, ResourceResponse};
use chrono::{TimeZone, Utc};
use hmac::{Hmac, Mac};
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;

fn query(params: &BTreeMap<String, String>) -> String {
    let mut values = params.iter().map(|(key, value)| (crate::rpc_encode(key), crate::rpc_encode(value))).collect::<Vec<_>>();
    values.sort();
    values.into_iter().map(|(key, value)| format!("{key}={value}")).collect::<Vec<_>>().join("&")
}

fn sign(key: &[u8], value: &str) -> Result<Vec<u8>, String> {
    let mut mac: Hmac<Sha256> = <Hmac<Sha256> as Mac>::new_from_slice(key).map_err(|error| error.to_string())?;
    mac.update(value.as_bytes());
    Ok(mac.finalize().into_bytes().to_vec())
}

async fn request(
    service: &str, version: &str, action: &str, mut params: BTreeMap<String, String>, region: &str,
    access_key_id: &str, access_key_secret: &str,
) -> Result<Value, String> {
    let host = "open.volcengineapi.com";
    let region = if region.is_empty() { "cn-beijing" } else { region };
    params.insert("Action".into(), action.into());
    params.insert("Version".into(), version.into());
    let query = query(&params);
    let datetime = Utc::now().format("%Y%m%dT%H%M%SZ").to_string();
    let date = &datetime[..8];
    let payload_hash = format!("{:x}", Sha256::digest(b""));
    let canonical_headers = format!("x-date:{datetime}\n");
    let signed_headers = "x-date";
    let canonical_request = format!("GET\n/\n{query}\n{canonical_headers}\n{signed_headers}\n{payload_hash}");
    let credential_scope = format!("{date}/{region}/{service}/request");
    let string_to_sign = format!("HMAC-SHA256\n{datetime}\n{credential_scope}\n{:x}", Sha256::digest(canonical_request.as_bytes()));
    let date_key = sign(access_key_secret.as_bytes(), date)?;
    let region_key = sign(&date_key, region)?;
    let service_key = sign(&region_key, service)?;
    let signing_key = sign(&service_key, "request")?;
    let signature = hex::encode(sign(&signing_key, &string_to_sign)?);
    let authorization = format!("HMAC-SHA256 Credential={access_key_id}/{credential_scope}, SignedHeaders={signed_headers}, Signature={signature}");
    let response = match reqwest::Client::new().get(format!("https://{host}/?{query}"))
        .header("X-Date", &datetime).header("Authorization", authorization)
        .timeout(std::time::Duration::from_secs(25)).send().await {
        Ok(response) => response,
        Err(error) => {
            let message = format!("火山引擎请求失败: {error}");
            write_api_log(access_key_id, host, action, &json!(params), None, "失败", Some(&message));
            return Err(message);
        }
    };
    let status = response.status();
    let data: Value = match response.json().await {
        Ok(data) => data,
        Err(error) => {
            let message = format!("火山引擎返回解析失败: {error}");
            write_api_log(access_key_id, host, action, &json!(params), None, "失败", Some(&message));
            return Err(message);
        }
    };
    if !status.is_success() || data.pointer("/ResponseMetadata/Error").is_some() || data.get("Error").is_some() {
        let message = data.pointer("/ResponseMetadata/Error/Message").and_then(Value::as_str)
            .or_else(|| data.pointer("/ResponseMetadata/Error/Code").and_then(Value::as_str))
            .or_else(|| data.pointer("/Error/Message").and_then(Value::as_str))
            .or_else(|| data.get("Message").and_then(Value::as_str)).unwrap_or("火山引擎 API 返回错误");
        write_api_log(access_key_id, host, action, &json!(params), Some(&data), "失败", Some(message));
        return Err(message.into());
    }
    write_api_log(access_key_id, host, action, &json!(params), Some(&data), "成功", None);
    Ok(data.get("Result").cloned().unwrap_or(data))
}

async fn json_request(
    service: &str, version: &str, action: &str, payload: Value, region: &str,
    access_key_id: &str, access_key_secret: &str,
) -> Result<Value, String> {
    let region = if region.is_empty() { "cn-beijing" } else { region };
    let host = format!("{service}.volcengineapi.com");
    let mut params = BTreeMap::new();
    params.insert("Action".into(), action.into());
    params.insert("Version".into(), version.into());
    let query = query(&params);
    let body = serde_json::to_string(&payload).map_err(|error| format!("火山引擎请求序列化失败: {error}"))?;
    let datetime = Utc::now().format("%Y%m%dT%H%M%SZ").to_string();
    let date = &datetime[..8];
    let payload_hash = format!("{:x}", Sha256::digest(body.as_bytes()));
    let canonical_headers = format!("x-date:{datetime}\n");
    let signed_headers = "x-date";
    let canonical_request = format!("POST\n/\n{query}\n{canonical_headers}\n{signed_headers}\n{payload_hash}");
    let credential_scope = format!("{date}/{region}/{service}/request");
    let string_to_sign = format!("HMAC-SHA256\n{datetime}\n{credential_scope}\n{:x}", Sha256::digest(canonical_request.as_bytes()));
    let date_key = sign(access_key_secret.as_bytes(), date)?;
    let region_key = sign(&date_key, region)?;
    let service_key = sign(&region_key, service)?;
    let signing_key = sign(&service_key, "request")?;
    let signature = hex::encode(sign(&signing_key, &string_to_sign)?);
    let authorization = format!("HMAC-SHA256 Credential={access_key_id}/{credential_scope}, SignedHeaders={signed_headers}, Signature={signature}");
    let response = reqwest::Client::new().post(format!("https://{host}/?{query}"))
        .header("Accept", "application/json").header("Content-Type", "application/json").header("X-Date", &datetime).header("Authorization", authorization)
        .body(body).timeout(std::time::Duration::from_secs(25)).send().await
        .map_err(|error| format!("火山引擎请求失败: {error}"))?;
    let status = response.status();
    let text = response.text().await.map_err(|error| format!("火山引擎返回读取失败: {error}"))?;
    let data: Value = serde_json::from_str(&text).unwrap_or_else(|_| json!({"Message": text}));
    if !status.is_success() || data.pointer("/ResponseMetadata/Error").is_some() || data.get("Error").is_some() {
        let message = data.pointer("/ResponseMetadata/Error/Message").and_then(Value::as_str)
            .or_else(|| data.pointer("/ResponseMetadata/Error/Code").and_then(Value::as_str))
            .or_else(|| data.pointer("/Error/Message").and_then(Value::as_str))
            .or_else(|| data.get("Message").and_then(Value::as_str)).unwrap_or("火山引擎 API 返回错误");
        return Err(message.into());
    }
    Ok(data.get("Result").cloned().unwrap_or(data))
}

async fn tos_list_buckets(region: &str, access_key_id: &str, access_key_secret: &str) -> Result<Vec<Value>, String> {
    let region = if region.is_empty() { "cn-beijing" } else { region };
    let host = format!("tos-{region}.volces.com");
    let datetime = Utc::now().format("%Y%m%dT%H%M%SZ").to_string();
    let date = &datetime[..8];
    let payload_hash = format!("{:x}", Sha256::digest(b""));
    let canonical_headers = format!("host:{host}\nx-tos-content-sha256:{payload_hash}\nx-tos-date:{datetime}\n");
    let signed_headers = "host;x-tos-content-sha256;x-tos-date";
    let canonical_request = format!("GET\n/\n\n{canonical_headers}\n{signed_headers}\n{payload_hash}");
    let credential_scope = format!("{date}/{region}/tos/request");
    let string_to_sign = format!("TOS4-HMAC-SHA256\n{datetime}\n{credential_scope}\n{:x}", Sha256::digest(canonical_request.as_bytes()));
    let date_key = sign(access_key_secret.as_bytes(), date)?;
    let region_key = sign(&date_key, region)?;
    let service_key = sign(&region_key, "tos")?;
    let signing_key = sign(&service_key, "request")?;
    let signature = hex::encode(sign(&signing_key, &string_to_sign)?);
    let authorization = format!("TOS4-HMAC-SHA256 Credential={access_key_id}/{credential_scope}, SignedHeaders={signed_headers}, Signature={signature}");
    let response = reqwest::Client::new().get(format!("https://{host}/"))
        .header("Host", &host).header("X-Tos-Date", &datetime).header("X-Tos-Content-Sha256", &payload_hash).header("Authorization", authorization)
        .timeout(std::time::Duration::from_secs(25)).send().await.map_err(|error| format!("TOS 请求失败: {error}"))?;
    let status = response.status();
    let body = response.text().await.map_err(|error| format!("TOS 返回读取失败: {error}"))?;
    if !status.is_success() {
        let message = xml_text(&body, "Message");
        return Err(format!("TOS 返回错误（{status}）：{}", if message.is_empty() { xml_text(&body, "Code") } else { message }));
    }
    let buckets = match serde_json::from_str::<Value>(&body) {
        Ok(data) => data.get("Buckets").and_then(Value::as_array).cloned()
            .or_else(|| data.pointer("/Buckets/Bucket").and_then(Value::as_array).cloned())
            .or_else(|| data.get("Bucket").and_then(Value::as_array).cloned())
            .unwrap_or_default(),
        Err(_) => xml_blocks(&body, "Bucket").into_iter().map(|bucket| json!({"Name": xml_text(&bucket, "Name"), "Location": xml_text(&bucket, "Location"), "CreationDate": xml_text(&bucket, "CreationDate")})).collect(),
    };
    Ok(buckets.into_iter().map(|bucket| {
        let name = bucket.get("Name").or_else(|| bucket.get("BucketName")).and_then(Value::as_str).unwrap_or("");
        let location = bucket.get("Location").or_else(|| bucket.get("Region")).and_then(Value::as_str).filter(|value| !value.is_empty()).unwrap_or(region);
        let creation_date = bucket.get("CreationDate").and_then(Value::as_str).unwrap_or("");
        json!({"Name": name, "Location": location, "CreationDate": creation_date, "StorageClass": "Standard", "ExtranetEndpoint": format!("{name}.tos-{location}.volces.com"), "Acl": "private"})
    }).filter(|bucket| bucket.get("Name").and_then(Value::as_str).is_some_and(|name| !name.is_empty())).collect())
}

fn instance(item: &Value, region: &str) -> Value {
    let mut value = item.clone();
    if let Value::Object(ref mut target) = value {
        let instance_id = item.get("InstanceId").or_else(|| item.get("InstanceID")).cloned().unwrap_or(json!(""));
        let status = item.get("Status").or_else(|| item.get("InstanceStatus")).cloned().unwrap_or(json!(""));
        target.insert("InstanceId".into(), instance_id.clone());
        target.insert("InstanceName".into(), item.get("InstanceName").cloned().unwrap_or(instance_id));
        target.insert("Status".into(), status.clone()); target.insert("InstanceStatus".into(), status);
        target.insert("PublicIpAddress".into(), item.get("PublicIpAddress").or_else(|| item.pointer("/PublicIpAddresses/0")).or_else(|| item.get("EipAddress")).cloned().unwrap_or(json!("")));
        target.insert("_region_id".into(), json!(region));
    }
    value
}

fn rds_instance(item: &Value, region: &str) -> Value {
    let mut value = item.clone();
    if let Value::Object(ref mut target) = value {
        let instance_id = item.get("DBInstanceId").or_else(|| item.get("InstanceId")).or_else(|| item.get("InstanceID")).cloned().unwrap_or(json!(""));
        target.insert("DBInstanceId".into(), instance_id.clone());
        target.insert("DBInstanceDescription".into(), item.get("DBInstanceName").or_else(|| item.get("InstanceName")).cloned().unwrap_or(instance_id));
        target.insert("DBInstanceStatus".into(), item.get("Status").or_else(|| item.get("DBInstanceStatus")).cloned().unwrap_or(json!("")));
        target.insert("Engine".into(), item.get("Engine").cloned().unwrap_or(json!("MySQL")));
        target.insert("_region_id".into(), json!(region));
    }
    value
}

fn swas_instance(item: &Value, region: &str) -> Value {
    let mut value = item.clone();
    if let Value::Object(ref mut target) = value {
        let instance_id = item.get("InstanceId").or_else(|| item.get("InstanceID")).cloned().unwrap_or(json!(""));
        let status = item.get("Status").or_else(|| item.get("InstanceStatus")).cloned().unwrap_or(json!(""));
        target.insert("InstanceId".into(), instance_id.clone());
        target.insert("InstanceName".into(), item.get("InstanceName").or_else(|| item.get("Name")).cloned().unwrap_or(instance_id));
        target.insert("Status".into(), status.clone()); target.insert("InstanceStatus".into(), status);
        target.insert("PublicIpAddress".into(), item.get("PublicIpAddress").or_else(|| item.get("PublicIp")).or_else(|| item.pointer("/PublicIpAddresses/0")).or_else(|| item.get("EipAddress")).cloned().unwrap_or(json!("")));
        target.insert("_region_id".into(), json!(region));
    }
    value
}

fn redis_instance(item: &Value, region: &str) -> Value {
    let mut value = item.clone();
    if let Value::Object(ref mut target) = value {
        let instance_id = item.get("InstanceId").or_else(|| item.get("InstanceID")).or_else(|| item.get("DBInstanceId")).or_else(|| item.get("RedisInstanceId")).cloned().unwrap_or(json!(""));
        let status = item.get("Status").or_else(|| item.get("InstanceStatus")).cloned().unwrap_or(json!(""));
        target.insert("KVStoreInstanceId".into(), instance_id.clone()); target.insert("InstanceId".into(), instance_id.clone());
        target.insert("InstanceName".into(), item.get("InstanceName").or_else(|| item.get("DBInstanceName")).or_else(|| item.get("Name")).cloned().unwrap_or(instance_id));
        target.insert("InstanceStatus".into(), status.clone()); target.insert("DBInstanceStatus".into(), status);
        target.insert("EngineVersion".into(), item.get("EngineVersion").or_else(|| item.get("RedisVersion")).cloned().unwrap_or(json!("Redis")));
        target.insert("_region_id".into(), json!(region));
    }
    value
}

fn edge_domain(item: &Value) -> Value {
    let mut value = item.clone();
    if let Value::Object(ref mut target) = value {
        let domain = item.get("DomainName").or_else(|| item.get("Domain")).or_else(|| item.get("Name")).cloned().unwrap_or(json!(""));
        target.insert("SiteId".into(), item.get("DomainId").or_else(|| item.get("DomainID")).cloned().unwrap_or(domain.clone()));
        target.insert("SiteName".into(), domain.clone()); target.insert("DomainName".into(), domain);
        target.insert("Status".into(), item.get("Status").or_else(|| item.get("DomainStatus")).cloned().unwrap_or(json!("")));
        target.insert("AccessType".into(), item.get("ServiceType").or_else(|| item.get("BusinessType")).cloned().unwrap_or(json!("CDN")));
        target.insert("Coverage".into(), item.get("Area").or_else(|| item.get("Scope")).cloned().unwrap_or(json!("")));
        target.insert("PlanName".into(), item.get("Plan").or_else(|| item.get("ProductType")).cloned().unwrap_or(json!("")));
    }
    value
}

fn dns_zone(item: &Value) -> Value {
    let mut value = item.clone();
    if let Value::Object(ref mut target) = value {
        let expires_at = item.get("ExpiredTime").and_then(Value::as_i64).unwrap_or_default();
        let expiration = if expires_at > 0 {
            let timestamp = if expires_at < 1_000_000_000_000 { expires_at.saturating_mul(1_000) } else { expires_at };
            Utc.timestamp_millis_opt(timestamp).single().map(|value| value.to_rfc3339()).unwrap_or_default()
        } else { String::new() };
        target.insert("DomainName".into(), item.get("ZoneName").cloned().unwrap_or(json!("")));
        target.insert("DomainStatus".into(), json!("正常"));
        target.insert("RegistrationDate".into(), item.get("CreatedAt").cloned().unwrap_or(json!("")));
        target.insert("ExpirationDate".into(), json!(expiration));
        target.insert("RecordCount".into(), item.get("RecordCount").cloned().unwrap_or(json!(0)));
    }
    value
}

pub(crate) async fn resource_items(id: i64, resource_type: &str, access_key_id: &str, access_key_secret: &str) -> ResourceResponse {
    let mut items = Vec::new();
    let mut errors = Vec::new();
    let region = match volc_region_id(id) {
        Ok(region) => region,
        Err(error) => return ResourceResponse { resource_type: resource_type.into(), items, errors: vec![error], fetched_at: Utc::now().timestamp_millis() },
    };
    match resource_type {
        "ecs" => match request("ecs", "2020-04-01", "DescribeInstances", string_params(&[("PageSize", "100".into()), ("PageNumber", "1".into())]), &region, access_key_id, access_key_secret).await {
            Ok(data) => { let source = if !array_at(&data, &["Instances"]).is_empty() { array_at(&data, &["Instances"]) } else { array_at(&data, &["Instances", "Instance"]) }; items.extend(source.into_iter().map(|item| instance(item, &region))); },
            Err(error) => errors.push(error),
        },
        "oss" => match tos_list_buckets(&region, access_key_id, access_key_secret).await { Ok(values) => items.extend(values), Err(error) => errors.push(error) },
        "domain" => match json_request("dns", "2018-08-01", "ListZones", json!({"PageSize": 100, "PageNumber": 1}), &region, access_key_id, access_key_secret).await {
            Ok(data) => items.extend(array_at(&data, &["Zones"]).into_iter().map(dns_zone)),
            Err(error) => errors.push(error),
        },
        "rds" => match request("rds_mysql", "2018-01-01", "DescribeDBInstances", string_params(&[("PageSize", "100".into()), ("PageNumber", "1".into())]), &region, access_key_id, access_key_secret).await {
            Ok(data) => { let source = if !array_at(&data, &["DBInstances"]).is_empty() { array_at(&data, &["DBInstances"]) } else { array_at(&data, &["Items"]) }; items.extend(source.into_iter().map(|item| rds_instance(item, &region))); },
            Err(error) => errors.push(error),
        },
        "swas" => match request("lighthouse", "2020-04-01", "DescribeInstances", string_params(&[("PageSize", "100".into()), ("PageNumber", "1".into())]), &region, access_key_id, access_key_secret).await {
            Ok(data) => { let source = if !array_at(&data, &["Instances"]).is_empty() { array_at(&data, &["Instances"]) } else { array_at(&data, &["InstanceSet"]) }; items.extend(source.into_iter().map(|item| swas_instance(item, &region))); },
            Err(error) => errors.push(error),
        },
        "redis" => match request("Redis", "2020-12-07", "DescribeDBInstances", string_params(&[("PageSize", "100".into()), ("PageNumber", "1".into())]), &region, access_key_id, access_key_secret).await {
            Ok(data) => { let source = if !array_at(&data, &["DBInstances"]).is_empty() { array_at(&data, &["DBInstances"]) } else { array_at(&data, &["Items"]) }; items.extend(source.into_iter().map(|item| redis_instance(item, &region))); },
            Err(error) => errors.push(error),
        },
        "esa" => match request("cdn", "2021-03-01", "ListCdnDomains", string_params(&[("PageSize", "100".into()), ("PageNumber", "1".into())]), &region, access_key_id, access_key_secret).await {
            Ok(data) => { let source = if !array_at(&data, &["Domains"]).is_empty() { array_at(&data, &["Domains"]) } else { array_at(&data, &["DomainList"]) }; items.extend(source.into_iter().map(edge_domain)); },
            Err(error) => errors.push(error),
        },
        other => errors.push(format!("火山引擎暂未接入 {other} 资源")),
    }
    ResourceResponse { resource_type: resource_type.into(), items, errors, fetched_at: Utc::now().timestamp_millis() }
}
